Fix guest customer incorrect subscription status - MAILPOET-5963

triple0t opened the pull request:
## Description

This PR updates the behavior of guest `checkout` with OptIn Unchecked.

Previously, when a Woo customer checks out without registration and leaves OptIn Unchecked, MailPoet creates a subscriber with the status `UNCONFIRMED` without any list.

These subscribers cannot be sent newsletters and are only counted toward the user's plan limits.

This is corrected in this PR ensuring the other features are working as expected.



## Code review notes

_N/A_

## QA notes

### Please test the following flows

#### Scenario 1 

* withWooCommerceCheckoutOptinEnabled
* withConfirmationEmailEnabled

**Test**: Check out a Woo product and leave optin **unchecked**.
-> A subscriber should be created with status `UNSUBSCRIBED`

**Test**: Check out a Woo product and leave optin **checked**.
-> A subscriber should be created with status `UNCONFIRMED`


#### Scenario 2
* withWooCommerceCheckoutOptinEnabled
* withConfirmationEmailDisabled

**Test**: Check out a Woo product and leave optin **unchecked**.
-> A subscriber should be created with status `UNSUBSCRIBED`

**Test**: Check out a Woo product and leave optin **checked**.
-> A subscriber should be created with status `SUBSCRIBED`


#### Scenario 3
* withWooCommerceCheckoutOptinDisabled
* withConfirmationEmailEnabled

**Test**: Check out a Woo product and leave optin **unchecked**.
-> A subscriber should be created with status `UNSUBSCRIBED`


#### Scenario 4
* withWooCommerceCheckoutOptinDisabled
* withConfirmationEmailDisabled

**Test**: Check out a Woo product and leave optin **unchecked**.
-> A subscriber should be created with status `UNSUBSCRIBED`
*** How was this even allowed?


## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-5963](https://mailpoet.atlassian.net/browse/MAILPOET-5963)

## After-merge notes

_N/A_

## Tasks

- [ ] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [ ] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5963]: https://mailpoet.atlassian.net/browse/MAILPOET-5963?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo commented on mailpoet/tests/integration/PostEditorBlocks/WooCommerceBlocksIntegrationTest.php:
```diff
@@ -60,21 +60,6 @@ public function testItHandlesOptInForGuestCustomer() {
-    $this->entityManager->refresh($subscriber);
-    verify($subscriber->getStatus())->equals(SubscriberEntity::STATUS_UNCONFIRMED);
-  }
-
```
I think it would be better rather to update the test to reflect the new behavior instead of removing it. We can test that the subscriber was not added. IMO, the test still has value. Some changes may cause us to start adding the subscriber, but this test will catch it.

triple0t replied:
Done.

costasovo commented on mailpoet/tests/acceptance/Subscriptions/WooCheckoutAutomateWooSubscriptionsCest.php:
```diff
@@ -55,18 +55,6 @@ public function checkoutOptInEnabled(\AcceptanceTester $i, $scenario) {
     $i->dontSee(self::AUTOMATE_WOO_OPTIN_TEXT);
   }
 
-  public function checkoutOptInUnchecked(\AcceptanceTester $i) {
```
Same as for the integration test, I think we should rather check the subscriber was not added instead of removing the tests.

triple0t replied:
Done

costasovo reviewed (changes requested):
@triple0t Thanks for jumping on fixing this issue!

This fixes the issue for sites that use blocks-based checkout pages, but the issue is still present for the classic checkout page, which is probably still used in many stores created before the blocks page was introduced.

### Replication of the issue on the classic checkout page

1) To setup the classic checkout page you can create a page and add shortcode `woocommerce_checkout`
2) After inserting the product to cart navigate to the page URL and you will see the classic checkout form
3) Fill in the form as a guest user and observe the issue

When I clicked `force sync` in the Lists > WooCommerce Customers the plugin synced the guest customer but this time with the expected `unsubscribed` status. I think this might be a bit confusing ("where they all suddenly came from?").

Considering the issue is also present on the classic checkout page and the force sync creates the subscriber anyway (I'm not sure we can avoid this) maybe we could change the logic so that the subscriber is still created but as as unsubscribed to get the same outcome as from the force sync. What do you think?

triple0t committed:
Set guest user by default to unsubscribed

The status will be updated when the user opts in for marketing email on checkout

MAILPOET-5963

triple0t committed:
Update tests to reflect the new logic

MAILPOET-5963

triple0t commented:
Thanks for the review, @costasovo 

The PR has been updated.

costasovo commented:
@triple0t Sorry for the late review. I missed the notification about your last comment, and I was not assigned to the PR, so I didn't notice I should re-review it. It is better to assign back to the reviewer after fixing so that the person who should act next is assigned.

Thank you for updating the tests. Did you address the issue with the classic checkout (see my comment about it: https://github.com/mailpoet/mailpoet/pull/5539#pullrequestreview-1988380227)? I don't see any changes related to this.

triple0t commented:
Hi @costasovo 👋🏾

> @triple0t Sorry for the late review. I missed the notification about your last comment, and I was not assigned to the PR, so I didn't notice I should re-review it. It is better to assign back to the reviewer after fixing so that the person who should act next is assigned.

Noted. Not a problem.
 
> Thank you for updating the tests. Did you address the issue with the classic checkout (see my comment about it: [#5539 (review)](https://github.com/mailpoet/mailpoet/pull/5539#pullrequestreview-1988380227))? I don't see any changes related to this.

Yup. 
Instead of creating the subscriber with status `UNCONFIRMED`, we now create them with the status `UNSUBSCRIBED`

This affects both Block and classic checkout.

When the user opts in for marketing emails, the status is updated back to `UNCONFIRMED`.

costasovo reviewed (approved):
> Instead of creating the subscriber with status UNCONFIRMED, we now create them with the status UNSUBSCRIBED

@triple0t Sorry I missed that both commit changed 🤦 It looks good! Moving it foeward!

Aschepikov reviewed (approved):
GJ!
